File: addons/terrabrush/Scripts/Tools/WaterTool.cs
#if TOOLS
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

public class WaterTool : ToolBase {
    private HashSet<ZoneResource> _sculptedZones;

    public WaterTool(TerraBrush terraBrush) : base(terraBrush) {}

    public override void BeginPaint() {
        base.BeginPaint();
        _sculptedZones = new HashSet<ZoneResource>();
    }

    public override void EndPaint() {
        base.EndPaint();

        _terraBrush.UpdateObjectsHeight(_sculptedZones.ToList());

        _sculptedZones = null;
    }

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        return zone.WaterTexture;
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        if (_terraBrush.WaterDefinition == null) {
            return;
        }

        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            var currentPixel = imageZoneInfo.Image.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);

            var newColor = toolType == TerrainToolType.WaterAdd ? new Color(1, currentPixel.G, currentPixel.B, 1) : new Color(0, 0.5f, 0.5f, 1);
            var newValue = new Color(
                Mathf.Lerp(currentPixel.R, newColor.R, pixelBrushStrength * brushStrength),
                Mathf.Lerp(currentPixel.G, newColor.G, pixelBrushStrength * brushStrength),
                Mathf.Lerp(currentPixel.B, newColor.B, pixelBrushStrength * brushStrength),
                1
            );
            imageZoneInfo.Image.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);

            _sculptedZones.Add(imageZoneInfo.Zone);
        });

        _terraBrush.TerrainZones.UpdateWaterTextures();
    }
}
#endif

File: src/editor_tools/water_tool.cpp
#include "water_tool.h"

using namespace godot;

void WaterTool::_bind_methods() {}

WaterTool::WaterTool() {}

WaterTool::~WaterTool() {}

bool WaterTool::getApplyResolution() const {
    return true;
}

void WaterTool::beginPaint() {
    ToolBase::beginPaint();

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();
}

void WaterTool::endPaint() {
    ToolBase::endPaint();

    TypedArray<Ref<ZoneResource>> sculptedZonesList = TypedArray<Ref<ZoneResource>>();
    for (Ref<ZoneResource> zone : _sculptedZones) {
        sculptedZonesList.append(zone);
    }
    _terraBrush->updateObjectsHeight(sculptedZonesList);

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();
}

Ref<Image> WaterTool::getToolCurrentImageTexture(Ref<ZoneResource> zone) {
    return zone->get_waterTexture();
}

void WaterTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    // If the water definition is null, or it's texture we exit here
    if (_terraBrush->get_waterDefinition().is_null() || _terraBrush->get_terrainZones()->get_zones().size() == 0 || Ref<ZoneResource>(_terraBrush->get_terrainZones()->get_zones()[0])->get_waterTexture().is_null()) {
        return;
    }

    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        Color currentPixel = imageZoneInfo.image->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);

        Color newColor = toolType == TerrainToolType::TERRAINTOOLTYPE_WATERADD ? Color(1, currentPixel.g, currentPixel.b, 1) : Color(0, 0.5f, 0.5f, 1);
        Color newValue = Color(
            Math::lerp(currentPixel.r, newColor.r, pixelBrushStrength * brushStrength),
            Math::lerp(currentPixel.g, newColor.g, pixelBrushStrength * brushStrength),
            Math::lerp(currentPixel.b, newColor.b, pixelBrushStrength * brushStrength),
            1
        );
        imageZoneInfo.image->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);

        _sculptedZones.insert(imageZoneInfo.zone);
    }));

    _terraBrush->get_terrainZones()->updateWaterTextures();
}
